### PR TITLE
Spline / Polynomial Geometry Refactoring

### PR DESCRIPTION
**Pull request by garethellis0**
### Description

 - Deleted `Polynomial` in `geom`
 - renamed `Polynomial` in `new_geom` to `Polynomial1d`
 - Created `Polynomial2d` as the combination of two `Polynomial1d`'s
 - Renamed `Spline` -> `LinearSpline` and moved `LinearSpline` from `geom` to `new_geom`
 - Created `Spline2d` and made `LinearSpline` inherit from it

### Testing Done
Unit tests.

### Resolved Issues
Part of #1382 
Potentially part of #1098, still a bit undecided on how I'm going to fit that in.

### Length Justification
Renamed a few things that expanded a scope, deleted several files.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/new_geom/polynomial1d.h
Comment (MathewMacDougall) [inline src/software/new_geom/polynomial1d.h]: Any reason not to use `Point` here? I'm guessing you may have done this to keep it more mathematical and keep away of spatial representations?

If we do keep it as points, may be convenient to have an easy to way to convert `Point` to `pair<double, double>` since I mostly see this getting used with points (not necessarily something for this PR).

File: src/software/new_geom/polynomial2d.cpp
Comment (MathewMacDougall) [inline src/software/new_geom/polynomial2d.cpp]: nit: missing newline at EOF

File: src/software/new_geom/polynomial2d.h
Comment (MathewMacDougall) [inline src/software/new_geom/polynomial2d.h]:
```suggestion
     * it always returns the point (0,0).
```

File: src/software/new_geom/polynomial2d_test.cpp
Comment (MathewMacDougall) [inline src/software/new_geom/polynomial2d_test.cpp]: should cover negative values too

File: src/software/new_geom/polynomial2d_test.cpp
Comment (MathewMacDougall) [inline src/software/new_geom/polynomial2d_test.cpp]: This is technically redundant with the tests for `get_poly_x` and `get_poly_y`

File: src/software/new_geom/spline2d.h
Comment (MathewMacDougall) [inline src/software/new_geom/spline2d.h]: nit: missing newline at EOF

File: src/software/new_geom/spline2d.h
Comment (MathewMacDougall) [inline src/software/new_geom/spline2d.h:20]: need javadocs for the public functions

File: src/software/new_geom/spline2d.h
Comment (MathewMacDougall) [inline src/software/new_geom/spline2d.h]: nit: define the order of spline segments. Eg from the start point to the end point

File: src/software/new_geom/linear_spline2d.h
Comment (MathewMacDougall) [inline src/software/new_geom/linear_spline2d.h:13]: Just curious, why not call this `FirstOrderSpline2d` or similar? Just asking because I expect that to be the naming scheme of higher-order splines, so might be worth being consistent.

File: src/software/new_geom/linear_spline2d.h
Comment (MathewMacDougall) [inline src/software/new_geom/linear_spline2d.h]: We already have the convention of not javadocing functions marked `override`, so I don't think we need these javadoc comments for these functions

File: src/software/new_geom/linear_spline2d.h
Comment (MathewMacDougall) [inline src/software/new_geom/linear_spline2d.h]: Don't understand the "time at index to index + 1" part of this comment. Not sure what time or index you are referring to.

File: src/software/new_geom/linear_spline2d.h
Comment (MathewMacDougall) [inline src/software/new_geom/linear_spline2d.h]: nit formatting: Add a period after the first sentence or split these lines with a blank line. Looks kinda like a weird run-on as it is right now

File: src/software/new_geom/linear_spline2d.cpp
Comment (MathewMacDougall) [inline src/software/new_geom/linear_spline2d.cpp:13]: nit: because the initializer list can be implicitly cast to a vector, why not put the implementation of `initLinearSegments` in the constructor that takes the vector, and then just call that constructor from here. 

Then don't have to remember to update initialization lists in two places / constructors if things change, and don't need the extra function.

File: src/software/new_geom/linear_spline2d.cpp
Comment (MathewMacDougall) [inline src/software/new_geom/linear_spline2d.cpp]:
```
* @return value of spline evaluated at value val. If val not in [0,1]
*         then the closest endpoint
```

This function does not properly implement the spec of the interface. It should clamp `val` to the range `[0, 1]`

File: src/software/new_geom/linear_spline2d.h
Comment (MathewMacDougall) [inline src/software/new_geom/linear_spline2d.h]:
```suggestion
     * @throws std::runtime_error if points.size() <= 1
```

File: src/software/new_geom/linear_spline2d.cpp
Comment (MathewMacDougall) [inline src/software/new_geom/linear_spline2d.cpp:76]: Not handling the case when `points.size() == 1`

File: src/software/new_geom/linear_spline2d.cpp
Comment (MathewMacDougall) [inline src/software/new_geom/linear_spline2d.cpp]: use `static_cast` as per our style guide

File: src/software/new_geom/linear_spline2d.h
Comment (MathewMacDougall) [inline src/software/new_geom/linear_spline2d.h]: `invalid_argument` seems like a more appropriate exception here?

File: src/software/new_geom/linear_spline_test.cpp
Comment (MathewMacDougall) [inline src/software/new_geom/linear_spline_test.cpp]: nit: technically checking `NumKnots` is out of scope of this unit test (according to the name). Consider breaking it up

File: src/software/new_geom/linear_spline_test.cpp
Comment (MathewMacDougall) [inline src/software/new_geom/linear_spline_test.cpp]: nit: technically checking `NumKnots` is out of scope of this unit test (according to the name). Consider breaking it up

File: src/software/new_geom/linear_spline_test.cpp
Comment (MathewMacDougall) [inline src/software/new_geom/linear_spline_test.cpp]: nit: technically checking `NumKnots` is out of scope of this unit test (according to the name). Consider breaking it up

File: src/software/new_geom/linear_spline_test.cpp
Comment (MathewMacDougall) [inline src/software/new_geom/linear_spline_test.cpp]: Consider separate unit tests for the start and end point functions (and numKnots as per my above comment)

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: src/software/new_geom/polynomial1d.h
Comment (jonathanlew) [reply src/software/new_geom/polynomial1d.h]: A constraint is a pair of (x,f(x)). For clarity, can we either use a new datatype to represent this or break it apart, i.e. `constructLinearPolynomialFromConstraints(double t1, double x_at_t1, double t2, double x_at_t2)`

File: src/software/new_geom/spline2d.h
Comment (jonathanlew) [inline src/software/new_geom/spline2d.h:62]: Since `SplineSegment2d` is the same for linear or higher order splines, `getValueAt`, `getKnots`, `getNumKnots`, etc. is the same no matter what order spline, so a derived class of `Spline2d` only needs to implement a different constructor. Can we implement these functions in Spline2d?

**Review by jonathanlew — CHANGES_REQUESTED** [review]
I like the overall structure of this. Just got some suggestions

File: src/software/new_geom/polynomial1d.h
Comment (garethellis0) [reply src/software/new_geom/polynomial1d.h]: Yeah, since it's a 1d polynomial a "Point" doesn't really make much sense. 

Changed to `double input_1, double output_1, double input_2, double output_2` to be super explict.

File: src/software/new_geom/polynomial2d.cpp
Comment (garethellis0) [reply src/software/new_geom/polynomial2d.cpp]: added

File: src/software/new_geom/polynomial2d.h
Comment (garethellis0) [reply src/software/new_geom/polynomial2d.h]: changed

File: src/software/new_geom/polynomial2d_test.cpp
Comment (garethellis0) [reply src/software/new_geom/polynomial2d_test.cpp]: done

File: src/software/new_geom/polynomial2d_test.cpp
Comment (garethellis0) [reply src/software/new_geom/polynomial2d_test.cpp]: removed these two checks

File: src/software/new_geom/spline2d.h
Comment (garethellis0) [reply src/software/new_geom/spline2d.h]: added

File: src/software/new_geom/spline2d.h
Comment (garethellis0) [reply src/software/new_geom/spline2d.h:20]: added

File: src/software/new_geom/spline2d.h
Comment (garethellis0) [reply src/software/new_geom/spline2d.h]: Added:`
```
     * @return The segments that make up this spline, with the order of segments being
     *         the order in which they appear along the spline, going from start to end.
```

File: src/software/new_geom/linear_spline2d.h
Comment (garethellis0) [reply src/software/new_geom/linear_spline2d.h:13]: I think `Linear`, `Quadratic`, `Cubic`, etc. is a more intuitive naming scheme for most people.

File: src/software/new_geom/linear_spline2d.h
Comment (garethellis0) [reply src/software/new_geom/linear_spline2d.h]: removed all

File: src/software/new_geom/linear_spline2d.h
Comment (garethellis0) [reply src/software/new_geom/linear_spline2d.h]: Yeah, this is a bit of a rubbish comment, not sure where it comes from.... changed to:
```
    // The segments making up this spline, from the start to the end
```

File: src/software/new_geom/linear_spline2d.h
Comment (garethellis0) [reply src/software/new_geom/linear_spline2d.h]: changed to:
```
     * Initialize segments with points.size() - 1 linear segments, interpolating
     * the given points
```

File: src/software/new_geom/linear_spline2d.cpp
Comment (garethellis0) [reply src/software/new_geom/linear_spline2d.cpp]: Updated the implementation, and added a test.

File: src/software/new_geom/linear_spline2d.cpp
Comment (garethellis0) [reply src/software/new_geom/linear_spline2d.cpp:13]: Yeah, so we can't do this because passing the initializer list to the constructor at that point is ambiguous, and the compiler will complain about a cycle.

File: src/software/new_geom/linear_spline2d.h
Comment (garethellis0) [reply src/software/new_geom/linear_spline2d.h]: changed

File: src/software/new_geom/linear_spline2d.cpp
Comment (garethellis0) [reply src/software/new_geom/linear_spline2d.cpp:76]: What do we need to do in that case? the `points` are already set in the constructor and there are no segments since we only have a single point.

File: src/software/new_geom/linear_spline2d.cpp
Comment (garethellis0) [reply src/software/new_geom/linear_spline2d.cpp]: Fixed, and also cast the numerators, which we really should have been doing already.

File: src/software/new_geom/linear_spline2d.h
Comment (garethellis0) [reply src/software/new_geom/linear_spline2d.h]: Agreed, changed

File: src/software/new_geom/linear_spline_test.cpp
Comment (garethellis0) [reply src/software/new_geom/linear_spline_test.cpp]: After offline discussion we decided to leave as is, but rename to indicate we're effectively testing the constructor.

File: src/software/new_geom/linear_spline_test.cpp
Comment (garethellis0) [reply src/software/new_geom/linear_spline_test.cpp]: see above comment

File: src/software/new_geom/spline2d.h
Comment (garethellis0) [reply src/software/new_geom/spline2d.h:62]: I'm going to hold off on this one if that's OK. I definitely noticed the same thing, but the next thing I'm going to do is implement a cubic spline, which should make it very clear what can actually be shared.

Comment (jonathanlew) [reply src/software/new_geom/spline2d.h:62]: OK just take note of it somewhere, if you could add it to an issue, that'd be best

File: src/software/new_geom/linear_spline2d.h
Comment (MathewMacDougall) [reply src/software/new_geom/linear_spline2d.h:13]: ah right those are also names. SGTM

File: src/software/new_geom/linear_spline2d.cpp
Comment (MathewMacDougall) [reply src/software/new_geom/linear_spline2d.cpp:76]: Oh I guess that is a valid spline, just confused me when I read it

**Review by MathewMacDougall — APPROVED** [review]
LGTM pending other comments

**Review by jonathanlew — APPROVED** [review]
LGTM!

**Review by akhilveeraghanta — APPROVED** [review]
lgtm :fire:

File: src/software/new_geom/spline2d.h
Comment (garethellis0) [reply src/software/new_geom/spline2d.h:62]: Ok, added it as an acceptance criteria under https://github.com/UBC-Thunderbots/Software/issues/1382, since this will really all be done together.

**Pull request merged**
